scripts/validate_rst_title_capitalization.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python
+"""
+Validate that the titles in the rst files follow the proper capitalization convention.
+
+Print the titles that do not follow the convention.
+
+Usage::
+./scripts/validate_rst_title_capitalization.py doc/source/development/contributing.rst
+./scripts/validate_rst_title_capitalization.py doc/source/
+
+"""
+import argparse
+import sys
+import re
+import os
+from typing import Tuple, Generator, List
+
+
+CAPITALIZATION_EXCEPTIONS = {
+    "pandas",
+    "Python",
+    "IPython",
+    "PyTables",
+    "Excel",
+    "JSON",
+    "HTML",
+    "SAS",
+    "SQL",
+    "BigQuery",
+    "STATA",
+    "Interval",
+    "PEP8",
+    "Period",
+    "Series",
+    "Index",
+    "DataFrame",
+    "C",
+    "Git",
+    "GitHub",
+    "NumPy",
+    "Apache",
+    "Arrow",
+    "Parquet",
+    "MultiIndex",
+    "NumFOCUS",
+    "sklearn",
+    "Docker",
+}
+
+CAP_EXCEPTIONS_DICT = {word.lower(): word for word in CAPITALIZATION_EXCEPTIONS}
+
+err_msg = "Heading capitalization formatted incorrectly. Please correctly capitalize"
+
+
+def correct_title_capitalization(title: str) -> str:
+    """
+    Algorithm to create the correct capitalization for a given title
+
+    Parameters
+    ----------
+    title : str
+        Heading string to correct
+
+    Returns
+    -------
+    correct_title : str
+        Correctly capitalized heading
+
+    """
+
+    correct_title: str = re.sub(r"^\W*", "", title).capitalize()
+
+    removed_https_title = re.sub(r"<https?:\/\/.*[\r\n]*>", "", correct_title)
+
+    word_list = re.split(r"\W", removed_https_title)
+
+    for word in word_list:
+        if word.lower() in CAP_EXCEPTIONS_DICT:
+            correct_title = re.sub(
+                rf"\b{word}\b", CAP_EXCEPTIONS_DICT[word.lower()], correct_title
+            )
+
+    return correct_title
+
+
+def find_titles(rst_file: str) -> Generator[Tuple[str, int], None, None]:
+    """
+    Algorithm to identify particular text that should be considered headings in an
+    RST file
+
+    See <https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html> for details
+    on what constitutes a string as a heading in RST
+
+    Parameters
+    ----------
+    rst_file : str
+        RST file to scan through for headings
+
+    Yields
+    -------
+    title : str
+        A heading found in the rst file
+
+    line_number : int
+        The corresponding line number of the heading
+
+    """
+
+    with open(rst_file, "r") as file_obj:
+        lines = file_obj.read().split("\n")
+
+    regex = {
+        "*": r"^(?:\*{1})*$",
+        "=": r"^(?:={1})*$",
+        "-": r"^(?:-{1})*$",
+        "^": r"^(?:\^{1})*$",
+        "~": r"^(?:~{1})*$",
+        "#": r"^(?:#{1})*$",
+        '"': r'^(?:"{1})*$',
+    }
+
+    table = str.maketrans("", "", "*`_")
+
+    for line_no in range(1, len(lines)):
```

Comment:
I think it's more readable to use `for i, line in enumerate(lines):` and then access the previous line with `lines[i - 1]`.

Using regex also is probably overcomplicating things. You can do something like:
```python
line_chars = set(line)
if len(line_chars) == 1 and line_chats.pop() in ('=', '*',...) and len(line) == len(line[i - 1]):
  ... 
```